lib/smack/Prelude.cpp

```diff
@@ -168,6 +168,44 @@ FuncDecl *builtinOp(std::string baseName, const Attr *attr,
 const std::vector<unsigned> IntOpGen::INTEGER_SIZES{
     1, 5, 6, 8, 16, 24, 32, 40, 48, 56, 64, 80, 88, 96, 128, 160, 256};
 
+const std::map<unsigned, std::string> IntOpGen::INT_LIMITS{
+    {0, "1"},
+    {1, "2"},
+    {4, "16"},
+    {5, "32"},
```

Comment:
You have 5 twice. Also, where are the first numbers of these pairs coming from? I see that they do not match `INTEGER_SIZES` above, and so I am curious how you picked them.
Also, why do you need 0?

Reply:
This table computes both `2^(i-1)` and `2^i` for `i` in `INTEGER_SIZES`, that's probably I got 2 `5`s in the keys.

Reply:
I don't get it. Integer sizes are `1, 5, 6, 8, 16, 24, 32, 40, 48, 56, 64, 80, 88, 96, 128, 160, 256`.
In this map though you are mapping things such as `0`, `4`, `7`, etc., which are not in integer sizes.
So what you wrote above "This table computes both 2^(i-1) and 2^i for i in INTEGER_SIZES" seems wrong to me, in particular the part "for i in INTEGER_SIZES".

Could you please clarify?

Reply:
This manual coding of strings is not great, because we cannot identify potential mistakes visually. Why don’t we just create this map programmatically? Looks like the formula is: compute the power of two whose exponent is given by the key, and convert to string.

Reply:
> This manual coding of strings is not great, because we cannot identify potential mistakes visually. Why don’t we just create this map programmatically? Looks like the formula is: compute the power of two whose exponent is given by the key, and convert to string.

There are some bit-widths that are too large for 64 bit. So we probably need arbitrary precison integer library unless we want to write a print function for large bit-widths ourselves. The former solution would require a new dependency.

Reply:
We can use this: https://llvm.org/doxygen/classllvm_1_1APInt.html. We use it in the integer overflow checker for this purpose.